pyvera/__init__.py

```diff
@@ -241,9 +242,30 @@ def get_changed_devices(self, timestamp):
         payload.update({
             'id': 'lu_sdata',
         })
-        result = self.data_request(payload, TIMEOUT*2).json()
+        
+        LOG.debug("get_changed_devices() requesting payload %s", str(payload))
+        r = self.data_request(payload, TIMEOUT*2)
+        r.raise_for_status()
+
+        # Unfortunately, not all JSON implementations return the well-known
+        # json.decoder.JSONDecodeError, so attempt to trap and raise what
+        # the poller expects.
+        try:
+            result = r.json()
+        except json.decoder.JSONDecodeError:
+            raise # pass it up
+        except Exception as ex:
```

Comment:
It's generally considered bad practice to catch all exceptions like this. Mainly because some of what you might catch isn't what you intend (eg this would even catch Python syntax errors!).

This has a very restricted scope - so isn't a disaster - but did you try to be specific and have problems?

Reply:
The specific problem I've run into is that the JSON module that comes bundled with Hass.io does not throw the (somewhat) standard json.decode.JSONDecodeError on parse errors. But I think I've got the preliminary checks prior to parsing tight enough that parse errors would truly be... ahem... exceptional. I left it in, being conservative on pyvera's stability side. Now, not every best or bad practice is best or bad in every circumstance, but I agree it's overly broad now that other work mitigates much of the risk on the fetches. I'll pull it out.

That does, however, introduce a risk from a different direction. After making a successful fetch of data, updates are sent to the "observers" through callbacks, and anything, including any exception, can be raised in those callbacks outside of pyvera's control, but would take pyvera's thread down as a whole. As currently structured, the callbacks are invoked within the same try block as the data fetches, so once the current broad exception is removed, those exceptions arising in a callback will stop the loop. So, I think some additional protection around the invocation of *each* callback is warranted, again for the longevity of the loop, but the question then is, what are you comfortable with? Since a callback can do anything, this is one of those places where I would argue that the widest possible exception trap is perhaps warranted. Again, this is a very limited scope.

What are your thoughts?

Reply:
Actually, I had this general exception in two places... in the get_changed, and in polling loop itself. My comments above apply to the polling loop. Sorry for the confusion.

Reply:
I don't really have strong views - except that I think it's important to trace unexpected errors. 

Had a terrible time debugging some code where someone had caught everything - and something deep in the code sometimes bumped into an undefined variable. The code caught this silently!

So fine to silently catch the things you know about - but lets trace the unexpected.